File: submissions/online-judges/project-euler/DSU.py
class DSU:
    def __init__(self, n):
        self.p = [i for i in range(n)]
        self.sz = [1 for i in range(n)]

    def fpar(self, n):
        if n != self.p[n]:
            self.p[n] = self.fpar(self.p[n])
        return self.p[n]

    def merge(self, a, b):
        a = self.fpar(a)
        b = self.fpar(b)
        if a == b:
            return False
        if self.sz[a] < self.sz[b]:
            a, b = b, a
        self.sz[a] += self.sz[b]
        self.p[b] = a
        return True

n = 6
dsu = DSU(n)

dsu.merge(0, 1)
dsu.merge(1, 2)
dsu.merge(3, 5)
for i in range(n):
    print(i, dsu.fpar(i))

File: templates/cpp/cheatsheet/data-structures/dsu.cpp
#include <bits/stdc++.h>

using namespace std;

struct DSU {
    int n;
    vector<int> par, sz;
    DSU(int _n) : n(_n) {
        par.resize(n + 1);
        iota(par.begin(), par.end(), 0);
        sz.assign(n + 1, 1);
    }
    int fpar(int x) {
        return (x == par[x] ? x : par[x] = fpar(par[x]));
    }
    void merge(int x, int y) {
        x = fpar(x);
        y = fpar(y);
        if (x == y) {
            return;
        }
        if (sz[x] < sz[y]) {
            swap(x, y);
        }
        par[y] = x;
        sz[x] += sz[y];
    }
};
